avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastDeserializerGenerator.java

```diff
@@ -310,12 +336,23 @@ private void processRecord(JVar recordSchemaVar, String recordName, final Schema
               recordSchemaVar.invoke("getField").arg(field.name()).invoke("schema"));
         }
         fieldReuseSupplier = () -> result.invoke("get").arg(JExpr.lit(readerFieldPos));
+        fieldsWrittenToTarget++;
       }
       if (SchemaAssistant.isComplexType(field.schema())) {
-        processComplexType(fieldSchemaVar, field.name(), field.schema(), readerFieldSchema, methodBody, action,
+        processComplexType(fieldSchemaVar, field.name(), field.schema(), readerFieldSchema, popMethodBody, action,
             putExpressionInRecord, fieldReuseSupplier);
       } else {
-        processSimpleType(field.schema(), readerFieldSchema, methodBody, action, putExpressionInRecord, fieldReuseSupplier);
+        processSimpleType(field.schema(), readerFieldSchema, popMethodBody, action, putExpressionInRecord, fieldReuseSupplier);
+      }
+
+      if (popMethod != null) {
+        for(Class<? extends Exception> e: schemaAssistant.getExceptionsFromStringable()) {
+          popMethod._throws(e);
+        }
+      }
+
+      if (isTopLevel && (fieldsWrittenToTarget == recordReaderSchema.getFields().size())) {
```

Comment:
I am not sure whether this will be good for all the use cases since this behavior is different from Vanilla Avro.
If the input is a stream, which contains multiple records, and stop early will corrupt the deserialization of the following record.

Reply:
how would you suggest to control this or any other parameter (like the split size) would it be by constructor or system property? currently I haven't seen much examples of configurables

Reply:
We should add tests for deserializing multiple sequential records from a single byte stream. This is a regular functionality of Avro and we do use it, so it's important that whatever efficiency improvement we build doesn't break anything. 

In this case, I don't think there should be any configs. Things must work out of the box no matter if we read just one record or multiple ones.

What you can do is to short-circuit your work as soon as possible, but still keep state which lets you know that if the user asks for one more record, then you still have this baggage from the end of the last record to parse and skip before starting to deserialize the next record.

Reply:
short circuit sounds very complicated, "marking" your position would be complicated and my bigger concern is that you need to remember the binary decoder but what do you do if you are given a different decoder in the next call? you might be in a multi-threaded situation. I think the caller must provide this on the deserialization command, any thoughts about that?

Reply:
For sequential deserialization, there are some objects that keep getting reused -- at the very least the stream and I think the decoder too? There's got to be a way to keep a position somewhere...

Reply:
yes but I think they are kept by the caller? I'm not familiar with the api, we call the "read" method directly with a binary decoder that we generate before hand

Reply:
The most important is that sequential deser of many records from the same stream keeps working, and we should add a test for that. Once we have a test, it'll be easy to validate if any given optimization can be adopted or not. If it breaks this use case, and we cannot find a way to make it work, then we should abandon it.